### DISCUSSION TITLE
Better display rejected data on the APP

### DISCUSSION THREAD

**Discussion by xu4wang** [directus/directus#11504] [Archive]
I am wondering is there a way to better display the item detail data when the data is not accepted. 

The use case is as below:

1.  A user create one item and submit it,
2. A directus hook listening the "item.create" event, and check if the request data is good to save, if there is any field with wrong data, it will be rejected.  
3. When the item is rejected, the user will make changes to every wrong fields,  and re-submit. 

Is there a way to display the rejected fields in an easy to find style(for example in red) so the user will be easier to find which fields need to be corrected.

![image](https://user-images.githubusercontent.com/311397/152957218-29312d57-fe5f-482a-a402-5f21f8e4fc67.png)

for example, in the above form, if the user input the price too high, the save will not be successful, and user need to adjust price and save again.

**Comment by azrikahar** [comment]
There are two possible ways, first one would be simpler, whereas second one should be exactly what you are looking for but slightly more lines of code. Though I'm not particularly sure if the second way here is entirely accurate or could be simpler so cc @rijkvanzanten @nickrum just in case. 😄 Wouldn't want to end up misleading other users here 😬

## First way: using InvalidPayloadException

### Code

```js
module.exports = function ({ filter }, { exceptions }) {
	const { InvalidPayloadException } = exceptions;

	filter('test.items.create', async (input) => {
		if (input.value < 50) {
			throw new InvalidPayloadException('Value must be greater than or equal to 50');
		}
		
		// do other valid logic here

		return input;
	});
};
```

### Result

![5Jup6NQGLs](https://user-images.githubusercontent.com/42867097/152963687-cbf739d0-be85-46c0-9014-4c1b3296f0d1.gif)

## Second way: using FailedValidationException

### Code

```js
const { validatePayload } = require('@directus/shared/utils');
const { flatten } = require('lodash');

module.exports = function ({ filter }, { exceptions }) {
	const { FailedValidationException } = exceptions;

	filter('test.items.create', async (input) => {
		// create a filter rule
		// Reference: https://docs.directus.io/configuration/filter-rules
		const validation = {
			value: {
				_gte: 50,
			},
		};

		// create an array of validation errors to throw to the App, if any
		const validationErrors = [];
		validationErrors.push(
			...flatten(
				validatePayload(validation, input).map((error) =>
					error.details.map((details) => new FailedValidationException(details))
				)
			)
		);
		if (validationErrors.length > 0) throw validationErrors;

		// do other valid logic here

		return input;
	});
};
```

### Result

![91pnKg2tRq](https://user-images.githubusercontent.com/42867097/152966513-4df23bec-aa55-45a1-8040-39d68f656c62.gif)

**Reply by xu4wang** [reply]
For the second approach, if there are many fields in the item, can I return a list of exceptions for all the fields(null for good one does not need exception)? 
or maybe a { field_name: exception} map for all the in-correct fields?

**Reply by azrikahar** [reply]
The validationsErrors should do that for you automatically, which means correct ones won't be highlighted if I understood correctly. Here's another example where I change the `validation` variable into a more complex filter rule:

```js
		const validation = {
			_and: [
				{
					value: {
						_gte: 50,
					},
				},
				{
					name: {
						_contains: 'abc',
					},
				},
				{
					description: {
						_contains: 'special',
					},
				},
			],
		};
```

And we can see when I enter `special` for description, it is not highlighted. Only when I modify it then it'll say it must contain `special`:

![nUUiBGPwSI](https://user-images.githubusercontent.com/42867097/152972206-ab9cde07-6581-4e9f-a6e5-3c8354b86e78.gif)

**Reply by xu4wang** [reply]
Thanks for the comprehensive answer! It's a tutorial for me. Even better than the official doc. :)

In my case, I will have to write my own logic for validation, instead of using the standard validatepayload. But as long as the app understands a list of exceptions, I should be able to achieve it.

**Reply by VaZark** [reply]
This should part of the documentation tbh. I'm sure lots of projects would like to add this to some input fields or another

**Reply by benhaynes** [reply]
CC @erondpowell 🙂

**Comment by xu4wang** [comment]
One more question, do you think it's possible for the below scenario?  It's related to the validation discussed above.

The user can input wrong data and the item will be in draft status by default.
In a later stage, when the user want to change the item status into "published" and submit again, we make the validation and inform the app to display the Exceptions returned.

**Reply by azrikahar** [reply]
I believe you should be able to do an IF ELSE check on the status so that it will only do the validation if the status is `published`.

**Reply by xu4wang** [reply]
OK Will try
